proto/insonmnia.go

```diff
@@ -0,0 +1,135 @@
+package sonm
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+	"strconv"
+	"strings"
+	"time"
+
+	ds "github.com/c2h5oh/datasize"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+func (m *Duration) Unwrap() time.Duration {
+	return time.Second * time.Duration(m.GetSeconds())
+}
+
+func (m *Duration) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var v string
+	if err := unmarshal(&v); err != nil {
+		return err
+	}
+
+	d, err := time.ParseDuration(v)
+	if err != nil {
+		return err
+	}
+
+	m.Seconds = int64(d.Truncate(time.Second).Seconds())
+	return nil
+}
+
+func (m *EthAddress) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var v string
+	if err := unmarshal(&v); err != nil {
+		return err
+	}
+
+	if !common.IsHexAddress(v) {
+		return errors.New("invalid ethereum address format")
+	}
+
+	m.Address = common.HexToAddress(v).Bytes()
+	return nil
+}
+
+func (m *DataSize) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var v string
+	if err := unmarshal(&v); err != nil {
+		return err
+	}
+
+	var bs ds.ByteSize
+	if err := bs.UnmarshalText([]byte(strings.ToLower(v))); err != nil {
+		return err
+	}
+
+	m.Size = bs.Bytes()
+	return nil
+}
+
+func (m *DataSizeRate) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var v string
+	if err := unmarshal(&v); err != nil {
+		return err
+	}
+
+	v = trimTimeRate(v)
+	var bs ds.ByteSize
+	if err := bs.UnmarshalText([]byte(v)); err != nil {
+		return err
+	}
+
+	m.Size = bs.Bytes()
+	return nil
+}
+
+func trimTimeRate(v string) string {
+	v = strings.ToLower(v)
+	v = strings.Trim(v, `/s`)
+	return v
+}
+
+func (m *Price) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var v string
+	if err := unmarshal(&v); err != nil {
+		return err
+	}
+
+	price, err := m.parse(v)
+	if err != nil {
+		return err
+	}
+
+	m.PerSecond = price
+	return nil
+}
+
+func (m *Price) parse(v string) (*BigInt, error) {
```

Comment:
strange method signature. Either it should be separate function returning Price, error, or it shoud modify price itself and return nothing